docs/Upgrading.md

```diff
@@ -20,7 +22,29 @@ This file documents important changes needed to upgrade your app's Shopify App v
 
 [Upgrading from `v8.6` to `v9.0.0`](#upgrading-from-v86-to-v900)
 
+# General Advice
+
+Although we strive to make upgrades as smooth as possible, some effort may be required to stay up to date with the latest changes to `shopify_app`.
+
+We strongly recommend you avoid 'monkeypatching' any existing code from `ShopifyApp`, e.g. by inheriting from `ShopifyApp` and then overriding particular methods. This can result in difficult upgrades. If your app does so, you will need carefully check the gem's internal changes when upgrading.
+
+If you need upgrade by more than one major version (e.g. from v18 to v20), we recommend doing so one by major version at a time. Deploy each into production to help to detect problems earlier.
+
+If you do run into issues, there are two useful techniques to apply: Adding log statements, and using an interactive debugger, such as `pry`.
+
+You can temporarily add log statements or debugger calls to the `shopify_app` or `shopify-api-ruby` gems:
+  * You can modify a gem using [`bundle open`](https://boringrails.com/tips/bundle-open-debug-gems)
+  * Alternatively, you can your modify your `Gemfile` to use local locally checked out gems with the the [`path` option](https://bundler.io/man/gemfile.5.html).
+* Note that if you make changes to a gem, you will need to restart the app for the changes to be applied.
+
+# Common Errors
+
+* `redirect_uri is not whitelisted` – you likely need to update the App Setup on https://partners.shopify.com/.
+* `This app can’t load due to an issue with browser cookies` - this can be caused by an infinite redirect due to coding error.
```

Comment:
If we proceed with https://github.com/Shopify/shopify_app/pull/1530 then it should hopefully reduce the need for this.